syntax = "proto3";

package currencyconverter;

// Service definition
service CurrencyConverter {
  // Convert from one currency to another
  rpc Convert (ConversionRequest) returns (ConversionResponse) {}
  
  // Get all available currencies
  rpc GetAvailableCurrencies (EmptyRequest) returns (CurrenciesResponse) {}
  
  // Get exchange rate between two currencies
  rpc GetExchangeRate (ExchangeRateRequest) returns (ExchangeRateResponse) {}
}

// Empty request for getting all currencies
message EmptyRequest {}

// Request for currency conversion
message ConversionRequest {
  string from_currency = 1;
  string to_currency = 2;
  double amount = 3;
}

// Response for currency conversion
message ConversionResponse {
  double converted_amount = 1;
  string from_currency = 2;
  string to_currency = 3;
  double exchange_rate = 4;
  bool success = 5;
  string error_message = 6;
}

// Request for getting exchange rate
message ExchangeRateRequest {
  string from_currency = 1;
  string to_currency = 2;
}

// Response for exchange rate
message ExchangeRateResponse {
  double exchange_rate = 1;
  bool success = 2;
  string error_message = 3;
}

// Response for available currencies
message CurrenciesResponse {
  repeated string currencies = 1;
  bool success = 2;
  string error_message = 3;
}